<html>
<head>
    <base href="http://appris.bioinfo.cnio.es/">
    <title>{APPRIS} - Annotating principal splice isoforms</title>
    <meta name="description" content="{APPRIS} - Annotating principal splice isoforms">
    <meta name="author" content="Jose Rodriguez, CNIO-INB">
    <style type="text/css">
        .firestar {
            color: rgb(244,169,39);
        }
        .matador3d {
            color: rgb(128,0,0);
        }
        .spade {
            color: rgb(164,189,83);
        }
        .corsair {
            color: rgb(4,95,180);
        }
        .thump {
            color: rgb(245,169,208);
        }
        .crash {
            color: rgb(153,153,102);
        }
    </style>
</head>

<body>
<!-- SPECIFIC section -->
<h2>Description</h2>
<p>
    The <strong>firestar track</strong> predicts functionally important residues based on the fireDB database. The track shows the PDB ligands detected by firestar for each transcript.
</p>

<p>
    firestar is a module that is part of APPRIS system to provide annotations of alternative splice isoforms and identify principal isoforms.
</p>

<!-- COMMON section -->
<h2>Procedures</h2>
<p>
    APPRIS is a system [1] that deploys a range of computational methods to provide annotations of alternative splice isoforms and identify principal isoforms for multiple species. These annotations are based on the modules of the APPRIS Database [2].
</p>
<ul>
    <li>Functionally important residues, <b>firestar</b></li>
    <li>Protein structural information, <b>Matador3D</b></li>
    <li>Presence of whole protein domains, <b>SPADE</b></li>
    <li>Conservation against vertebrates, <b>CORSAIR</b></li>
    <li>Presence of whole trans-membrane helices, <b>THUMP</b></li>
    <li>Prediction of signal peptide and sub-cellular location, <b>CRASH</b></li>
</ul>
<img src="img/APPRIS_analysis.png" width="50%" style="display: block; margin-left: auto; margin-right: auto">

<h4 class="firestar">firestar: Functionally Important Residues</h4>
<p><a href="http://firedb.bioinfo.cnio.es/Php/FireStar.php" target="_blank">firestar</a> predicts functionally important residues based on the <a href="http://firedb.bioinfo.cnio.es/Php/FireDB.php" target="_blank">fireDB</a> database. The predictions are based on the local evaluation of alignments between the query sequence and the structures with functional information that are stored in FireDB. The reliability of predictions is assessed with <a href="http://firedb.bioinfo.cnio.es/Php/square.php" target="_blank">SQUARE</a> and the functional information is highlighted along with a reliability score.</p>
<p>Functional residues are highly conserved, even across large evolutionary distances. Since these residues are unlikely to have arisen by chance we can also use this to help determine the principal isoform.</p>
<p>Variants that have "lost" conserved functional residues are unlikely to be the principal isoform.</p>

<h4 class="matador3d">Matador3D: Protein structural information</h4>
<p>Protein structural information is analysed with Matador3D. In practice variant sequences from the same gene are mapped onto 3D structures by running BLAST against the <a href="http://www.pdb.org/" target="_blank">PDB</a>.</p>
<p>Protein structure is much more conserved than sequence and isoforms with large inserts or deletions relative to homologous crystal structures are also not likely to be principal.</p>

<h4 class="spade">SPADE: Scanning Pfam Alignments for Damaged Entities</h4>
<p>Proteins are generally comprised of one or more functional regions commonly termed domains. Identifying the functional domains present in a variant can provide insights into the function and to determine the most likely principal isoform.</p>
<p>The presence of functional domains is analysed with <a href="http://pfam.sanger.ac.uk" target="_blank">Pfamscan</a>.</p>
<p>Transcripts that have "lost" conserved functional domains are unlikely to be principal isoforms.</p>

<h4 class="corsair">CORSAIR: Species conservation</h4>
<p>CORSAIR carries out <a href="http://blast.ncbi.nlm.nih.gov/" target="_blank">BLAST</a> searches against
    vertebrates to determine the most likely principal isoform. The method counts the number of species that align correctly and without gaps for each variant.</p>
<p>Transcripts that are conserved over greater evolutionary distances are more likely to be the main variant.</p>

<h4 class="thump">THUMP: Detecting reliable trans-membrane helices</h4>
<p>THUMP makes unanimous predictions of trans-membrane helices using three different methods: <a
        href="http://bioinf.cs.ucl.ac.uk/psipred" target="_blank">MEMSAT 3.0</a>, <a href="http://phobius.sbc.su.se/" target="_blank">Phobius</a>, and <a href="http://topcons.cbr.su.se/index.php?about=proprodiv" target="_blank">PRODIV</a>. A helix has to be predicted by all three methods to be considered reliable.</p><p>Transcripts that have "lost" trans-membrane helices are less likely to be the principal isoform.</p>

<h4 class="crash">CRASH: Signal Peptide and Mitochondrial Signal Sequences.</h4>
<p>The presence and location of signal peptides and cleavage sites in amino acid sequences are analysed with <a
        href="http://www.cbs.dtu.dk/services/SignalP/" target="_blank">SignalP</a> program. And
    <a href="http://www.cbs.dtu.dk/services/TargetP/" target="_blank">TargetP</a> predicts the sub-cellular location of eukaryotic proteins. CRASH uses a rule-based analysis of these two programs to select only reliable predictions.</p>

<h2>Validation</h2>
<p>
    The selection of principal isoforms coincides with the main isoform detected in proteomics experiments for almost <strong>99.5%</strong> of comparable genes, showing that APPRIS principal isoforms are a clear improvement on choosing the longest isoform as the main protein isoform in the cell [3].
</p>

<h2>Credits</h2>
<p>
    The data was generated by <a href="http://www.cnio.es/" target="_blank">Spanish National Cancer Research Centre (CNIO)</a> and <a href="http://www.inab.org/" target="_blank">Spanish Institute of Bioinformatics</a> using a computational pipeline developed by Michael Tress and Jose Manuel Rodriguez.
</p>
<p>
    APPRIS was developed within the <a href="http://www.gencodegenes.org/" target="_blank">GENCODE consortium</a> [4] to address the challenge of annotating alternative protein isoforms of the Human genome, but it has been also annotated for another species.
</p>
<h2>Contacts</h2>
<p>
    If you have questions or comments, please write to:
<ul>
    <li>APPRIS GMail, <a title="Email contact" href="mailto:appris.cnio@gmail.com" target="_blank">appris.cnio@gmail.com</a></li>
    <li>Jose Rodriguez, <a title="Email contact" href="mailto:jmrodriguez@cnio.es" target="_blank">jmrodriguez@cnio.es</a></li>
</ul>
</p>

<h2>Data Release Policy</h2>
<p>
    All data is freely available to the public.
</p>

<h2>References</h2>
<p>
    [1] <b>APPRIS WebServer and WebServices.</b></br>
    Rodriguez JM, Carro A, Valencia A, Tress ML.</br>
    Nucleic Acids Res. 2015 May 18. doi: <a href="http://dx.doi.org/10.1093/nar/gkv512" target="_blank">10.1093/nar/gkv512</a></br>
</p>
<p>
    [2] <b>APPRIS: annotation of principal and alternative splice isoforms.</b></br>
    Rodriguez JM, Maietta P, Ezkurdia I, Pietrelli A, Wesselink JJ, Lopez G, Valencia A and Tress ML.</br>
    Nucleic acids research 2013;41;Database issue;D110-7. doi: <a href="http://dx.doi.org/10.1093/nar/gks1058" target="_blank">10.1093/nar/gks1058</a></br>
</p>
<p>
    [3] <b>Most highly expressed protein-coding genes have a single dominant isoform.</b></br>
    Ezkurdia I, Rodriguez JM, Carrillo-de Santa Pau E, Vazquez J, Valencia A, and Tress ML.</br>
    Proteome Res. 2015 Apr 3;14(4):1880-7. doi: <a href="http://dx.doi.org/10.1021/pr501286b" target="_blank">10.1021/pr501286b</a></br>
</p>
<p>
    [4] <b>GENCODE: the reference annotation for the ENCODE Project.</b></br>
    Harrow J., Frankish A., Gonzalez J.M., Tapanari E., Diekhans M., Kokocinski F., Aken B.L., Barrell D., Zadissa A., Searle S., et al.</br>
    Genome Res. 2012;22:1775-1789. doi: <a href="http://dx.doi.org/10.1101/gr.135350.111" target="_blank">10.1101/gr.135350.111</a></br>
</p>
</body>
</html>
