
<html>
<head>
    <title>{PROTEO} - Peptide evidence for the human genome</title>
    <meta name="description" content="Peptide evidence for the human genome">
    <meta name="author" content="Jose Rodriguez, CNIO-INB">
</head>
<style type="text/css">
table td {
	padding: 7px;
}
</style>
<body>
	<h2>Description</h3>
    <p>
        Peptides that were identified[1], in collaboration, by <a href="https://www.cnio.es/" target="_blank">Spanish National Cancer Research Centre (CNIO)</a> and <a href="https://www.cnic.es/" target="_blank">Centro Nacional de Investigaciones Cardiovasculares (CNIC)</a>.
    </p>
    <p>
    	We interrogate peptides from eight large-scale human proteomics experiments and databases, irrespective of tissue or cell type, for the vast majority of the protein-coding genes in these experiments. The peptide has to be included in at least 2 experiments.
	</p>
	
    <h2>Procedures</h2>
    <p>
        We collected the peptides for the analysis from eight distinct large-scale proteomics data sets: the <a href="http://www.peptideatlas.org/" target="_blank">PeptideAtlas</a>[2] and <a href="https://peptide.nist.gov/" target="_blank"> NIST databases</a> and six published large-scale experiments.[3-8]. For the Wilhelm analysis, we only included the peptides from the publically available Cellzome experiments on human tissues.
    </p>
    <p>
    	The eight studies covered a huge range of tissues and cell types: the peptides from the PeptideAtlas database cover 51 different tissues, cell types, and developmental stages, whereas the Geiger study interrogated 11 different cell types. The spectra from the PeptideAtlas database were only part of the NIST database and the Ezkurdia analyses. The Kim and Wilhelm analyses peptides were generated from 30 and 35 distinct tissues types (51 tissues in total).    	
	</p>
	
    <h2>Annotations</h2>
    <p>
    	The following table describes the annotations of PROTEO tracks:
    </p>    
	<table border="1" cellspacing="0">
	<tbody>
		<tr>
			<th>field</th>
			<th>example</th>
			<th>Info type</th>
			<th>description</th>
		</tr>
		<tr>
			<td><tt>Item</tt></td>
			<td>GVSEAAVLKPSEELPAEATSSVEPEK</td>
			<td><tt>string</tt></td>
			<td>Peptide found in 8 large-scale human proteomics experiments and databases</td>
		</tr>
		<tr>
			<td><tt>Score</tt></td>
			<td>2</td>
			<td><tt>int range [2-8]</tt></td>
			<td>Number experiments where the peptide were found. The peptide has to be included in at least 2 experiments</td>
		</tr>
		<tr>
			<td><tt>Position</tt></td>
			<td>chr21:33043745-33043822</td>
			<td width="120px"><tt>chr:start-end</tt></td>
			<td>Genome position</td>
		</tr>
		<tr>
			<td><tt>List of Ensembl transcript ID</tt></td>
			<td>ENST00000399804;ENST00000286835</td>
			<td><tt>string</tt></td>
			<td>List of variants (Ensembl transcript ID) that contains the peptide</td>
		</tr>
	</tbody>
	</table>
    	
    <h2>Credits</h2>
    <p>
        The data was generated by <a href="https://www.cnio.es/" target="_blank">Spanish National Cancer Research Centre (CNIO)</a>, <a href="https://www.cnic.es/" target="_blank">Centro Nacional de Investigaciones Cardiovasculares (CNIC)</a> and <a href="https://inb-elixir.es/" target="_blank">Spanish Institute of Bioinformatics</a> using a computational pipeline developed by Michael Tress, Iakes Ezkurdia and Jose Manuel Rodriguez.
    </p>

    <h2>Contacts</h2>
    <p>
        If you have questions or comments, please write to:
    <ul>
        <li>Jose Rodriguez, <a title="Email contact" href="mailto:jmrodriguez@cnio.es" target="_blank">jmrodriguez@cnio.es</a></li>
        <li>Michael
            Tress, <a title="Email contact" href="mailto:mtress@cnio.es" target="_blank">mtress@cnio.es</a></li>
    </ul>
    </p>

	
	<h2>Data Release Policy</h2>
    <p>
    	All data is freely available to the public.
	</p>
	
	<h2>References</h2>
    <p>
    	[1] Ezkurdia I, Rodriguez JM, Carrillo-de Santa Pau E, Vazquez J, Valencia A, Tress ML.<br>
    	<b>Most highly expressed protein-coding genes have a single dominant isoform.</b><br>
        J Proteome Res. 2015 Apr 3;14(4):1880-7. doi: <a href="https://doi.org/10.1021/pr501286b">10.1021/pr501286b</a>. Epub 2015 Mar 11. PubMed PMID: 25732134.
    </p>
		[2] Farrah, T.; Deutsch, E. W.; Hoopmann, M. R.; Hallows, J. L.; Sun, Z.; Huang, C. Y.; Moritz, R. L.<br>
    	The state of the human proteome in 2012 as viewed through PeptideAtlas.<br>
    	J. Proteome Res. 2013, 12, 162-171
	</p>
		[3] Ezkurdia, I.; del Pozo, A.; Frankish, A.; Rodriguez, J. M.; Harrow, J.; Ashman, K.; Valencia, A.; Tress, M. L.<br>
    	Comparative proteomics reveals a significant bias toward alternative protein isoforms with conserved structure and function.<br>
    	Mol. Biol. Evol. 2012, 29, 2265– 2283
	</p>
		[4] Munoz, J.; Low, T. Y.; Kok, Y. J.; Chin, A.; Frese, C. K.; Ding, V.; Choo, A.; Heck, A. J.<br>
		The quantitative proteomes of human-induced pluripotent stem cells and embryonic stem cells.<br>
		Mol. Syst. Biol. 2011, 7, 550
	</p>
		[5] Nagaraj, N.; Wisniewski, J. R.; Geiger, T.; Cox, J.; Kircher, M.; Kelso, J.; Pääbo, S.; Mann, M.<br>
		Deep proteome and transcriptome mapping of a human cancer cell line.<br>
		Mol. Syst. Biol. 2011, 7, 548
	</p>
		[6] Geiger, T.; Wehner, A.; Schaab, C.; Cox, J.; Mann, M.<br>
		Comparative proteomic analysis of eleven common cell lines reveals ubiquitous but varying expression of most proteins.<br>
		Mol. Cell Proteomics. 2012, 11, M111.014050
	</p>
		[7] Kim, M. S.; Pinto, S. M.; Getnet, D.; Nirujogi, R. S.; Manda, S. S.; Chaerkady, R.; Madugundu, A. K.; Kelkar, D. S.; Isserlin, R.; Jain, S.<br>
		A draft map of the human proteome.<br>
		Nature 2014, 509, 575-581
	</p>
		[8] Wilhelm, M.; Schlegl, J.; Hahne, H.; Moghaddas Gholami, A.; Lieberenz, M.; Savitski, M. M.; Ziegler, E.; Butzmann, L.; Gessulat, S.; Marx, H.<br>
		Mass-spectrometry-based draft of the human proteome.<br>
		Nature 2014, 509, 582-587
	</p>
	<p>
	Note: The <b>bold</b> title corresponds to the main paper that provides the collected peptides.
	</p>
</body>
</html>
