<!DOCTYPE html>
<html>
<head>
    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-128430091-1"></script>
    <script src="https://www.googleoptimize.com/optimize.js?id=GTM-N76Q8HW"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());
      gtag('config', 'UA-128430091-1', {'anonymize_ip': true, 'optimize_id': 'GTM-N76Q8HW'});
    </script>
    <!-- End Google Analytics -->
    <meta charset="UTF-8">
    <title>Important Alternative Isoforms!</title>
    <meta name="description" content="View important alternative isoforms based on APPRIS annotations of human genes in GENCODE release 27.">
    <!-- Imports CSS -->
    <link rel='stylesheet' type='text/css' href='css/reset.css'>
    <link rel="stylesheet" type='text/css' href="css/bootstrap.min.css">
    <link rel='stylesheet' type='text/css' href='css/appris.css'>

    <link rel="stylesheet" type='text/css' href="css/handsontable-7.4.0.full.min.css">
    <link rel="stylesheet" type='text/css' href="css/main.css">
    <link rel="stylesheet" type='text/css' href="css/list.css">

    <!-- Imports JS -->
    <script type='text/javascript' src="js/jquery-3.5.1.min.js"></script>
    <script type='text/javascript' src="js/bootstrap.min.js"></script>
    <script type='text/javascript' src="js/handsontable-7.4.0.full.min.js"></script>
    <script type='text/javascript' src="js/xlsx.full.min.js"></script>
    <script type='text/javascript' src="js/main.js"></script>


</head>
<body class="swagger-section">
<div class="row clearfix">
    <nav class="navbar navbar-default navbar-fixed-top navbar-appris" role="navigation">
        <div class="navbar-header">
            <button aria-expanded="false" class="navbar-toggle collapsed"
                    data-target="#appris-navbar-collapse" data-toggle="collapse"
                    type="button">
                <span class="sr-only">
                    Toggle navigation
                </span>
                <span class="icon-bar">
                </span>
                <span class="icon-bar">
                </span>
                <span class="icon-bar">
                </span>
            </button>
            <a class="navbar-brand" href="https://appris.bioinfo.cnio.es/">
                {APPRIS}
            </a>
        </div>

        <div class="collapse navbar-collapse" id="appris-navbar-collapse">
            <ul class="nav navbar-nav">
                <li class="dropdown" dropdown is-open="status.isopen">
                    <a href="https://appris.bioinfo.cnio.es/#/tools" class="dropdown-toggle" dropdown-toggle ng-disabled="disabled">Tools <b class="caret"></b></a>
                    <ul class="dropdown-menu">
                        <li><a href="https://appris.bioinfo.cnio.es/#/seeker">Database</a></li>
                        <li><a href="https://appris.bioinfo.cnio.es/#/server">WebServer</a></li>
                        <li><a href="https://apprisws.bioinfo.cnio.es/apidoc/">WebServices</a></li>
                    </ul>
                </li>
                <li>
                    <a href="https://appris.bioinfo.cnio.es/#/downloads">Downloads</a>
                </li>
    	        <li class="dropdown" dropdown is-open="status.isopen">
        	    <a class="dropdown-toggle btn" href="https://apprisws.bioinfo.cnio.es/appris_datasets/" dropdown-toggle ng-disabled="disabled">APPRIS Datasets<b class="caret"></b></a>
                    <ul class="dropdown-menu">
	               <li><a href="https://apprisws.bioinfo.cnio.es/appris_datasets/tissue_specific_alternative_splicing/">Tissue-Specific Alternative Splicing</a></li>
        	       <li><a href="https://apprisws.bioinfo.cnio.es/appris_datasets/tandem_exon_derived_substitutions/">Tandem Exon Derived Substitutions</a></li>
                 <li><a href="https://apprisws.bioinfo.cnio.es/appris_datasets/GENCODEv45-Refseq-UniProtKB_28_3_2024/">GENCODEv45-Refseq-UniProtKB_28_3_2024 human coding gene set</a></li>
                </ul>
                </li>
                <li class="dropdown" dropdown is-open="status.isopen">
                  <a class="dropdown-toggle btn" href="https://apprisws.bioinfo.cnio.es/appris_mane_matches/" dropdown-toggle ng-disabled="disabled">APPRIS/MANE matches<b class="caret"></b></a>
                    <ul class="dropdown-menu">
                      <li><a href="https://apprisws.bioinfo.cnio.es/appris_mane_matches/gencode19/">Gencode 19</a></li>
                     <li><a href="https://apprisws.bioinfo.cnio.es/appris_mane_matches/gencode40/">Gencode 40</a></li>
                     <li><a href="https://apprisws.bioinfo.cnio.es/appris_mane_matches/gencode41/">Gencode 41</a></li>
                     <li><a href="https://apprisws.bioinfo.cnio.es/appris_mane_matches/gencode42/">Gencode 42</a></li>
                     <li><a href="https://apprisws.bioinfo.cnio.es/appris_mane_matches/gencode43/">Gencode 43</a></li>
                     <li><a href="https://apprisws.bioinfo.cnio.es/appris_mane_matches/gencode44/">Gencode 44</a></li>
                     <li><a href="https://apprisws.bioinfo.cnio.es/appris_mane_matches/gencode45/">Gencode 45</a></li>
                     <li><a href="https://apprisws.bioinfo.cnio.es/appris_mane_matches/gencode46/">Gencode 46</a></li>
                     <li><a href="https://apprisws.bioinfo.cnio.es/appris_mane_matches/gencode47/">Gencode 47</a></li>
                     <li><a href="https://apprisws.bioinfo.cnio.es/appris_mane_matches/gencode48/">Gencode 48</a></li>
                     <li><a href="https://apprisws.bioinfo.cnio.es/appris_mane_matches/refseq105/">RefSeq 105</a></li>
                     <li><a href="https://apprisws.bioinfo.cnio.es/appris_mane_matches/refseq110/">RefSeq 110</a></li>
                    </ul>
                </li>
		            <li>
                    <a href="https://appris.bioinfo.cnio.es/#/help/intro">Help & Docs</a>
                </li>
                <li>
                    <a href="https://appris.bioinfo.cnio.es/#/about">About us</a>
                </li>
            </ul>
        </div>
    </nav>
</div>
<div>
  <header>
    <h1>APPRIS Datasets</h1>
  </header>
<div>
  <header>
    <h2>Important Isoforms</h2>
  </header>
  <p>
    As part of the work with APPRIS, we are investigating patterns in alternative splicing 
    at the protein level. Since the proteins are the functional end product of coding genes, 
    we believe that it is important to look at trends at the protein level, rather than the 
    transcript level, where a certain amount of biological and technical noise can often 
    obscure any signal. There is little cost associated with this noise at the transcript level, 
    but large amounts of misfolded proteins are potentially damaging at the protein level, and 
    will be controlled at some level, though we do not yet know the extent of this control, 
    or the processes behind it.
  </p>
  <p>
    However, we do know that there are numerous examples of functional alternative isoforms, and 
    as part of our work, we will attempt to list as many as we can. List of functional isoforms that 
    we find will be linked from this page when we find time to add them. The isoforms are also 
    listed in the individual papers.
  </p>
  <p>
    Note that the pairs of transcripts listed in these files are tied to a single release of 
    Ensembl/GENCODE. The sequences or stable IDs might change in subsequent versions. Transcripts 
    may appear in more than one list.
  </p>

  <p>
    Functional isoforms lists we have gathered so far:
  </p>
  
  <h4><strong><a href="tissue_specific_alternative_splicing/">· Tissue-Specific Alternative Splicing</a></strong></h4>
  <p>These isoforms come from this paper on tissue-specific alternative splicing:
    <address>
      <strong>An analysis of tissue-specific alternative splicing at the protein level</strong><br>
      Jose Manuel Rodriguez, Fernando Pozo, Tomas di Domenico, Jesus Vazquez, Michael L. Tress<br>
      DOI:<a href="https://doi.org/10.1371/journal.pcbi.1008287" target="_blank">10.1371/journal.pcbi.1008287</a>
    </address>
  </p>
  
  <hr/>

  <h4><strong><a href="tandem_exon_derived_substitutions/">· Tandem Exon Derived Substitutions</a></strong></h4>
  <p>Tandem exon derived substitutions involve the swapping of one (or more) coding exons for homologous coding exons. 
    These may be at the 5' or 3' end, or be produced by mutually exclusive splicing. These transcripts produce two 
    homologous proteins (or more) from the same gene. There are two papers detailing these isoforms: 
    <address>
      <strong>The clinical importance of tandem exon duplication-derived substitutions</strong><br>
      Laura Martinez Gomez , Fernando Pozo , Thomas A Walsh , Federico Abascal , Michael L Tress<br>
      DOI:<a href="https://doi.org/10.1093/nar/gkab623" target="_blank">10.1093/nar/gkab623</a>
    </address>
    <address>
      <strong>Origins and Evolution of Human Tandem Duplicated Exon Substitution Events</strong><br>
      Laura Martinez-Gomez , Daniel Cerdán-Vélez , Federico Abascal , Michael L Tress<br>
      DOI:<a href="https://doi.org/10.1093/gbe/evac162" target="_blank">10.1093/gbe/evac162</a>
    </address>
  </p>
	</div>

  <hr/>

  <div>
    <header>
      <h2>Reference gene sets</h2>
    </header>
    <p>
      We have merged the Ensembl/GENCODE, RefSeq and UniProtKB human reference gene sets. The initial 
      merge was automatic based on gene name, but the merge was finalised by several rounds of manual 
      curatcion. Genes that are not present in all three datasets or that are tagged as potential 
      no-coding genes may not be real coding genes.
    </p>
    <p>
      This merge was carried out with the GENCODE v45, RefSeq 110 and UniProtKB 28/3/2024 versions. 
      Details can be found in the paper.
    </p>
    
    <h4><strong><a href="GENCODEv45-Refseq-UniProtKB_28_3_2024/">· GENCODEv45-Refseq-UniProtKB_28_3_2024 human coding gene set</a></strong></h4>
  
    </div>
    <hr/>

 <footer></footer>

</body>
</html>
