<div>
    <p>The “APPRIS annotations” section shows the scores of the <a href="#/help/methods">APPRIS modules</a>.</p>
    <p>
        <img src="img/report-annots.png" class="img-thumbnail img-responsive img-center" style="width: 100%">
    </p>
    <h4>No. Functional Residues</h4>
    <p>firestar predicts functionally important residues. The box shows the total number of functional residues detected by firestar for each transcript.</p>

    <h4>Tertiary Structure Score</h4>
    <p>Matador3D (and Matador3Dv2) analyze protein structural information for each isoform. For
        both, protein sequences from the same gene are mapped onto 3D structures by running BLAST
        against the <a href="https://www.wwpdb.org/" target="_blank">PDB</a>, but depending on
        whether or not we have genomic information, the score is calculated in one of two ways.
    </p>
    <p>When we made use of genomic coordinates, the score is based on the number of exons that can be mapped
        to structural homologues. When we don't know the genomic coordinates, the score is the sum of bit-scores in
        PDB alignment.
    </p>
    <p>This score is only comparable for transcripts within the same gene and should not be used to compare
        between genes.</p>

    <h4>Conservation score (vertebrates)</h4>
    <p>CORSAIR carries out BLAST searches against vertebrates to determine the most likely principal isoform. The score approximates to the number of species that have isoforms that can be aligned to each tested isoform without introducing gaps.</p>

    <h4>Domains score</h4>
    <p>SPADE identifies the functional domains present in a transcript and detects those domains that are damaged
        (not whole). The number represents the sum of bitscores in Pfam alignment.</p>

    <h4>No. Transmembrane Helices</h4>
    <p>THUMP makes unanimous predictions of trans-membrane helices. The score shows the number of trans-membrane helices predicted by THUMP for each variant.</p>

    <h4>Signal Peptide Sequences and Mitochondrial Signal Sequences</h4>
    <p>CRASH predicts the presence and location of signal peptides and cleavage sites in amino acid sequences using an especially reliable form of SignalP and TargetP. Isoforms that are predicted to have reliable signal sequences are marked.</p>

    <h4>No. Mapping Peptides</h4>
    <p>Proteomic evidence only for the human genome (GENCODE gene set).</p>
    <p>PROTEO incorporates proteomic evidence from various mass spectrometry
       (MS) sources, covering a range of tissues and cell types.</p>
    <p>Prior to GENCODE v33, peptides were collected from 8 sources
       (<a href="https://doi.org/10.1021/pr501286b" target="_blank">Ezkurdia et al. 2015</a>;
       <a href="https://doi.org/10.1021/pr301012j" target="_blank">Farrah et al. 2013</a>;
       <a href="https://doi.org/10.1093/molbev/mss100" target="_blank">Ezkurdia et al. 2012</a>;
       <a href="https://doi.org/10.1038/msb.2011.84" target="_blank">Munoz et al. 2011</a>;
       <a href="https://doi.org/10.1038/msb.2011.81" target="_blank">Nagaraj et al. 2011</a>;
       <a href="https://doi.org/10.1074/mcp.m111.014050" target="_blank">Geiger et al. 2012</a>;
       <a href="https://doi.org/10.1038/nature13302" target="_blank">Kim et al. 2014</a>;
       <a href="https://doi.org/10.1038/nature13319" target="_blank">Wilhelm et al. 2014</a>).
       To improve reliability, peptides from each of these sources were filtered, eliminating
       non-tryptic and semi-tryptic peptides and peptides containing missed cleavages, and
       where possible only considering peptides identified by multiple search engines.
    </p>
    <p>APPRIS releases from GENCODE v33 onwards have used peptide evidence from 2 proteomics studies
       (<a href="https://doi.org/10.1038/nature13302" target="_blank">Kim et al. 2014</a>;
       <a href="https://doi.org/10.15252/msb.20188503" target="_blank">Wang et al. 2019</a>),
       and from GENCODE v38, these were supplemented by peptides from a further 5 studies
       (<a href="https://doi.org/10.1038/nature13438" target="_blank">Zhang et al. 2014</a>;
       <a href="https://doi.org/10.1016/j.cels.2017.05.009" target="_blank">Bekker-Jensen et al. 2017</a>;
       <a href="https://doi.org/10.1038/s41593-017-0011-2" target="_blank">Carlyle et al. 2017</a>;
       <a href="https://doi.org/10.1074/mcp.RA118.001170" target="_blank">Schiza et al. 2019</a>;
       <a href="https://doi.org/10.1016/j.cell.2020.08.036" target="_blank">Jiang et al. 2020</a>).
       For these studies, the Comet search engine
       (<a href="https://doi.org/10.1002/pmic.201200439" target="_blank">Eng et al. 2012</a>)
       was used with default parameters, then post-processed with Percolator
       (<a href="https://doi.org/10.1038/nmeth1113" target="_blank">Käll et al. 2007</a>;
       <a href="https://pubmed.ncbi.nlm.nih.gov/27572102/" target="_blank">The et al. 2016</a>).
       Peptide-spectrum matches (PSMs) that had a Posterior Error
       Probability (PEP) of lower than 0.001 were allowed as long
       as they were fully tryptic peptides and had no more than 2
       missed cleavages. Only peptides with at least 2 valid
       PSMs across the 7 studies are considered by PROTEO.
    </p>
    <h4>Normalized Functional Importance</h4>
    <p>TRIFID is a machine learning classifier based on APPRIS and external database input which we have found to be a reliable indicator of isoform functional importance.</p>
    <p>The normalized TRIFID score is shown if TRIFID predictions are available for the given isoform.</p>
    <p>For more information about this method, consult the <a href="https://gitlab.com/bu_cnio/trifid">TRIFID GitLab repository</a>.</p>

</div>
