<div>
    <p>The “APPRIS annotations” section shows the scores of the <a href="#/help/methods">APPRIS modules</a>.</p>
    <p>
        <img src="img/report-annots.png" class="img-thumbnail img-responsive img-center" style="width: 100%">
    </p>
    <h4>No. Functional Residues</h4>
    <p>firestar predicts functionally important residues. The box shows the total number of functional residues detected by firestar for each transcript.</p>

    <h4>Tertiary Structure Score</h4>
    <p>Matador3D (and Matador3Dv2) analyze protein structural information for each isoform. For
        both, protein sequences from the same gene are mapped onto 3D structures by running BLAST
        against the <a href="https://www.wwpdb.org/" target="_blank">PDB</a>, but depending on
        whether or not we have genomic information, the score is calculated in one of two ways.
    </p>
    <p>When we made use of genomic coordinates, the score is based on the number of exons that can be mapped
        to structural homologues. When we don't know the genomic coordinates, the score is the sum of bit-scores in
        PDB alignment.
    </p>
    <p>This score is only comparable for transcripts within the same gene and should not be used to compare
        between genes.</p>

    <h4>Conservation score (vertebrates)</h4>
    <p>CORSAIR carries out BLAST searches against vertebrates to determine the most likely principal isoform. The score approximates to the number of species that have isoforms that can be aligned to each tested isoform without introducing gaps.</p>

    <h4>Domains score</h4>
    <p>SPADE identifies the functional domains present in a transcript and detects those domains that are damaged
        (not whole). The number represents the sum of bitscores in Pfam alignment.</p>

    <h4>No. Transmembrane Helices</h4>
    <p>THUMP makes unanimous predictions of trans-membrane helices. The score shows the number of trans-membrane helices predicted by THUMP for each variant.</p>

    <h4>Signal Peptide Sequences and Mitochondrial Signal Sequences</h4>
    <p>CRASH predicts the presence and location of signal peptides and cleavage sites in amino acid sequences using an especially reliable form of SignalP and TargetP. Isoforms that are predicted to have reliable signal sequences are marked.</p>

    <h4>No. Mapping Peptides</h4>
    <p>Proteomic evidence only for the human genome (GENCODE gene set).</p>
    <p>PROTEO collects peptides from seven separate MS sources. Two came from large-scale proteomics
        databases, <a href="http://www.peptideatlas.org/">PeptideAtlas</a> and
        <a href="https://peptide.nist.gov/">NIST</a>. Another four datasets were recently published
        large-scale MS experiments. For all six datasets the starting point was the list of peptides
        provided by the authors or databases. We generated the final set of peptides (referred to as
        “CNIO") in house from an <a href="https://www.ncbi.nlm.nih.gov/pubmed/14558131">X!Tandem</a>
        search against spectra from the <a
        href="https://www.ncbi.nlm.nih.gov/pubmed/15595733">GPM</a> and PeptideAtlas databases,
        following the protocol set out in <a
        href="https://pubmed.ncbi.nlm.nih.gov/22446687">Ezkurdia et al</a> with a false discovery
        rate of 0.1%. These seven studies cover a wide range of search engines, tissues and cell
        types.
    </p>
    <p>In order to improve reliability the peptides from each of these studies were filtered, eliminating non-tryptic and semi-tryptic peptides and peptides containing missed cleavages. For those studies where it was possible we considered only peptides identified by multiple search engines.</p>

    <h4>Normalized Functional Importance</h4>
    <p>TRIFID is a machine learning classifier based on APPRIS and external database input which we have found to be a reliable indicator of isoform functional importance.</p>
    <p>The normalized TRIFID score is shown if TRIFID predictions are available for the given isoform.</p>
    <p>For more information about this method, consult the <a href="https://gitlab.com/bu_cnio/trifid">TRIFID GitLab repository</a>.</p>

</div>
